File: src/Systems/Objective-C/proxy.m
#import <Foundation/Foundation.h>

@protocol DocumentStore <NSObject>
- (NSString *)documentWithId:(NSInteger)documentId;
@end

@interface RemoteDocumentStore : NSObject <DocumentStore>
@property(nonatomic) NSInteger fetches;
@end

@implementation RemoteDocumentStore
- (NSString *)documentWithId:(NSInteger)documentId {
    self.fetches += 1;
    return [NSString stringWithFormat:@"doc(%ld)", (long)documentId];
}
@end

@interface DocumentStoreProxy : NSObject <DocumentStore>
@property(nonatomic, strong) RemoteDocumentStore *backend;
@property(nonatomic, strong) NSMutableDictionary<NSNumber *, NSString *> *cache;
@property(nonatomic) NSInteger backendCreations;
@end

@implementation DocumentStoreProxy
- (instancetype)init {
    self = [super init];
    if (self) {
        _cache = [NSMutableDictionary dictionary];
    }
    return self;
}

- (NSString *)documentWithId:(NSInteger)documentId {
    NSNumber *key = @(documentId);
    NSString *cached = self.cache[key];
    if (cached != nil) {
        return cached;
    }
    if (self.backend == nil) {
        self.backend = [[RemoteDocumentStore alloc] init];
        self.backendCreations += 1;
    }
    NSString *value = [self.backend documentWithId:documentId];
    self.cache[key] = value;
    return value;
}
@end

int main(void) {
    @autoreleasepool {
        DocumentStoreProxy *store = [[DocumentStoreProxy alloc] init];
        NSString *first = [store documentWithId:42];
        NSString *second = [store documentWithId:42];
        printf("backend=%ld;fetches=%ld;first=%s;second=%s\n",
               (long)store.backendCreations,
               (long)store.backend.fetches,
               first.UTF8String,
               second.UTF8String);
    }
    return 0;
}

File: src/Scripting/PythonPY/proxy.py
class RemoteDocumentStore:
    def __init__(self) -> None:
        self.fetch_count = 0

    def get(self, document_id: int) -> str:
        self.fetch_count += 1
        return f"doc({document_id})"


class DocumentStoreProxy:
    def __init__(self) -> None:
        self._backend: RemoteDocumentStore | None = None
        self._cache: dict[int, str] = {}

    @property
    def backend_count(self) -> int:
        return 0 if self._backend is None else 1

    @property
    def fetch_count(self) -> int:
        return 0 if self._backend is None else self._backend.fetch_count

    def get(self, document_id: int) -> str:
        if document_id in self._cache:
            return self._cache[document_id]
        if self._backend is None:
            self._backend = RemoteDocumentStore()
        value = self._backend.get(document_id)
        self._cache[document_id] = value
        return value


store = DocumentStoreProxy()
first = store.get(42)
second = store.get(42)
print(f"backend={store.backend_count};fetches={store.fetch_count};first={first};second={second}")
